packages/block-editor/src/components/link-control/index.js

```diff
@@ -0,0 +1,219 @@
+	// State
+	const [ inputValue, setInputValue ] = useState( '' );
+	const [ isEditingLink, setIsEditingLink ] = useState( true );
+
+	// Refs
+	const autocompleteRef = useRef( null );
+
+	// Handlers
+	const onInputChange = ( value = '' ) => {
+		setInputValue( value );
+	};
+
+	const closeLinkUI = () => {
+		resetInput();
+	};
+
+	const resetInput = useCallback( () => {
+		setInputValue( '' );
+	} );
+
+	const onLinkSelect = ( event, suggestion ) => {
+		event.preventDefault();
+		event.stopPropagation();
+
+		setIsEditingLink( false );
+
+		if ( isFunction( onLinkChange ) ) {
+			onLinkChange( suggestion );
+		}
+	};
+
+	const onStartEditing = () => {
+		setIsEditingLink( true );
+	};
+
+	const handleURLSearch = async ( value ) => {
+		return [ {
+			id: 1,
+			title: value,
+			type: 'URL',
+			url: value,
+		} ];
+	};
+
+	const stopPropagation = ( event ) => {
+		event.stopPropagation();
+	};
+
+	const stopPropagationRelevantKeys = ( event ) => {
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
+			// Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
+			stopPropagation( event );
+		}
+	};
+
+	// Effects
+	const getSearchHandler = useCallback( ( value ) => {
+		return ( /^https?:/.test( value ) ) ? handleURLSearch( value ) : fetchSearchSuggestions( value );
+	}, [ handleURLSearch, fetchSearchSuggestions ] );
+
+	// Render Components
+	const renderSearchResults = ( { suggestionsListProps, buildSuggestionItemProps, suggestions, selectedSuggestion } ) => {
+		/* eslint-disable react/jsx-key */
+		return (
+			<div className="block-editor-link-control__search-results-wrapper">
+				<div { ...suggestionsListProps } className="block-editor-link-control__search-results">
+					{ suggestions.map( ( suggestion, index ) => (
+						<button
+							{ ...buildSuggestionItemProps( suggestion, index ) }
+							onClick={ partialRight( onLinkSelect, suggestion ) }
+							className={ classnames( 'block-editor-link-control__search-item', {
+								'is-selected': index === selectedSuggestion,
+							} ) }
+
+						>
+							{ suggestion.type.toLowerCase() === 'url' && (
+								<Icon className="block-editor-link-control__search-item-icon" icon="admin-site-alt3" />
+							) }
+							<span className="block-editor-link-control__search-item-header">
+								<span className="block-editor-link-control__search-item-title">
+									<TextHighlight text={ suggestion.title } highlight={ inputValue } />
+								</span>
+								<span className="block-editor-link-control__search-item-info">{ suggestion.info || filterURLForDisplay( safeDecodeURI( suggestion.url ) ) || '' }</span>
+							</span>
+							<span className="block-editor-link-control__search-item-type">{ suggestion.type.toLowerCase() || '' }</span>
+						</button>
+					) ) }
+				</div>
+			</div>
+		);
+		/* eslint-enable react/jsx-key */
+	};
+
+	const LinkControlAdditionalSettings = () => (
```

Comment:
On a performance related note, my understanding is that components shouldn't be declared inline like this.

For function components, react will check whether this is the same component being rendered using a simple equality check. Since this is a different function, it'll be considered a different component, so React will replace the DOM element it renders each time the `LinkControl` component is rendered.

Reply:
Good point. We'll have to work through optimisations once the implementation is solid, but certainly one to keep a note of. I'll add to `Todos`.

Reply:
Now this is it's own component.